src/function/scalar/list/list_intersect.cpp

```diff
@@ -0,0 +1,258 @@
+			const idx_t it_idx = iter_list.offset + j;
+			const idx_t it_entry = iter_fmt.sel->get_index(it_idx);
+			if (!iter_fmt.validity.RowIsValid(it_entry)) {
+				continue;
+			}
+
+			const auto &key = iter_keys[it_entry];
+			if (set.find(key) == set.end() || result_set.find(key) != result_set.end()) {
+				continue;
+			}
+			result_set.insert(key);
+
+			const idx_t emit_left_idx = use_l_for_hash ? key_to_index_map[key] : it_idx;
+
+			result_sel.set_index(offset + row_result_length, emit_left_idx);
+			row_result_length++;
+		}
+
+		result_data[i].length = row_result_length;
+		offset += row_result_length;
+	}
+
+	ListVector::SetListSize(result, offset);
+
+	// If types differ, we need to slice into a temporary vector first, then cast
+	if (original_left_child_type != current_left_child_type) {
+		// Slice into a temporary vector with the coerced type
+		Vector temp_result_entry(current_left_child_type, offset);
+		temp_result_entry.Slice(l_child, result_sel, offset);
+		temp_result_entry.Flatten(offset);
+		FlatVector::SetValidity(temp_result_entry, result_entry_validity_mask);
+
+		// Cast the temporary vector to the original left child type
+		string error_message;
+		if (VectorOperations::DefaultTryCast(temp_result_entry, result_entry, offset, &error_message, false)) {
+			// Cast succeeded, result_entry now has the correct type
+		} else {
+			// Cast failed, fall back to slicing directly (this shouldn't happen if types are compatible)
+			result_entry.Slice(l_child, result_sel, offset);
+			result_entry.Flatten(offset);
+			FlatVector::SetValidity(result_entry, result_entry_validity_mask);
+		}
+	} else {
+		// Types match, slice directly
+		result_entry.Slice(l_child, result_sel, offset);
+		result_entry.Flatten(offset);
+		FlatVector::SetValidity(result_entry, result_entry_validity_mask);
+	}
+
+	result.SetVectorType(args.AllConstant() ? VectorType::CONSTANT_VECTOR : VectorType::FLAT_VECTOR);
+}
+
+static unique_ptr<FunctionData> ListIntersectBind(ClientContext &context, ScalarFunction &bound_function,
+                                                  vector<unique_ptr<Expression>> &arguments) {
+	D_ASSERT(bound_function.arguments.size() == 2);
+	arguments[0] = BoundCastExpression::AddArrayCastToList(context, std::move(arguments[0]));
+	arguments[1] = BoundCastExpression::AddArrayCastToList(context, std::move(arguments[1]));
+
+	// Handle NULL case
+	if (arguments[0]->return_type == LogicalType::SQLNULL) {
+		bound_function.return_type = LogicalType::SQLNULL;
+		return make_uniq<ListIntersectBindData>(LogicalType::SQLNULL);
+	}
+
+	// Store the original left child type before any type coercion happens
+	// This allows us to preserve the left list's element type in the result
+	auto original_left_child_type = ListType::GetChildType(arguments[0]->return_type);
+
+	// Set the return type to preserve the left list's element type
+	bound_function.return_type = LogicalType::LIST(original_left_child_type);
```

Comment:
Type inference (in this case template resolution) will happen before any function bind callbacks are called, so Im not sure if you even need ListIntersectBind at all